rfc-173-testing-our-work.md

```diff
@@ -0,0 +1,81 @@
+process was introduced relatively recently and there's no good reason why our
+testing choices haven't been added so far. This proposal aims to fill this gap
+and officially document our stance on testing as well as provide some sensible
+defaults.
+
+## Problem
+
+- Testing patterns - the subject of writing tests is a broad one and there are
+  numerous ways we can choose to test our work. Without guiding standards teams
+  can spend unnecessary time at the start of each project deciding their
+  approach, this time could be better spent elsewhere. With diverging approaches
+  the team can be less clear on how to test. These differences introduce
+  avoidable friction when people switch between projects
+- Tooling divergence - without testing guidelines we risk introducing multiple
+  tools that perform the same function. Whilst
+  [we are empowered to diverge from our defaults](https://github.com/dxw/tech-team-rfcs#4-standards-are-defaults)
+  there should be a good reason when doing so. Again, this can take time to set
+  up for new projects and make it harder for people to move between projects and
+  support
+- Team growth - the development team is growing and we often invite contractors
+  to represent us. Having an agreed set of principles will help all new joiners
+  build an understanding around what we expect from one another when testing
+- Stack variety - dxw are increasingly working with different stacks. These
+  often require us to use different or new testing tools. Having a way to record
+  our preferences as we grow will make it easier to contribute or work on those
+  stacks again in future
+
+## Proposal
+
+The topic of testing code has many facets. The aim of this proposal is to
+capture some of the higher level principles we wish to follow but separate from
+the details of how we achieve them. We propose to document those finer details
+in our
+[standards for testing](https://docs.google.com/document/d/1pSZkrviFAiR-0U-PGixOyC2MfiBdCIWIN66YyLa23nc/edit#heading=h.owf78h5xtccv)
+that may change more frequently. These standards may eventually become a guide
+in [the dxw Playbook](https://github.com/dxw/playbook).
+
+- We MUST write automated tests for all the code we write that will (or could)
+  be used by people in live environments – this includes internal dxw tools and
+  tooling that provisions real infrastructure
+- We MUST write appropriate test coverage for the code we write inline with our
+  standards for testing. This includes using
+  [the right type of test](https://thoughtbot.com/blog/rails-test-types-and-the-testing-pyramid)
+  as well as setting up and working with appropriate testing boundaries
+- We MUST configure our continuous integration tools to automatically run our
+  tests as part of our deployment pipelines. It MUST prevent deployments
+  completing should any test fail
+- We MUST set up new projects to automatically require 100% code coverage
```

Comment:
Do we/should we *require* 100% code coverage? Code coverage tools can be helpful guidance/nudge, but I don't think this should be a must.

One reason is I'm keen to ensure that we don't have RFCs that already don't apply to our existing projects - does everything we own have 100% code coverage? Are we going to go back and fix those that don't?

Reply:
👋 To my knowledge we have never strictly required 100% test coverage at dxw. We have tended to write lots of tests and then sometimes added monitoring for it later and were then unable to get it back to 100%.

Here I'm proposing that only _new_ projects aim for 100% allowing old ones to stay as they are. I quite agree that going back to apply this change over past work isn't feasible. 

Code coverage feedback is a useful as a nudge but from what I've seen there are problems that undermine it when it's not set to 100%:

Untested code gets through CI and slowly builds up quietly until it trips the threshold. The team then face a coverage report that says there are x number of LOC that have no coverage. This puts the unlucky person in a difficult position of having to retrospectively make it 100% or perhaps just fix one or two tests to make CI happy for the time being. The cycle then continues.

I've seen thresholds removed from 96 to 95% to get around this because 96 is totally arbitrary and doesn't mean anything.

I'm know I'm being bold by suggesting it become a MUST rather than a SHOULD but I think it's the right thing to do going forwards? If something can't be tested for a good reason the nudge could be to comment it out with nodoc or similar.

Reply:
Another way might be to have it set to 100% but not enforcing the test suite to hit it. My gut tells me the team would eventually become numb to the noisey nudges?

Reply:
I agree @tahb, I think set 100% as the default for new projects, and if a project has a good reason to change that then they can.

Reply:
Is it worth adding a change coverage requirement for old projects, eg "we SHOULD set up existing projects to require 100% coverage of code changes"? That way we aren't failing on things which are untested already, but we make sure new things get the attention.

Reply:
I'm happy with what we seem to have agreed on in these comments, which is that new projects should aim to have 100% test coverage unless they have explicit reasons not to. Still, I don't think that's what this will be read as without this chat thread and would be more comfortable with either more qualifiers or as a SHOULD as @jacksonj04 suggested.

Reply:
I'm putting "new" in bold and adding a SHOULD as suggested. Thanks for your contributions!

> - We MUST set up **new** projects to automatically require 100% code coverage
> - We SHOULD set up existing projects to require 100% code coverage